### PR TITLE
Nkey auth

### PR DESCRIPTION
**Pull request by aricart**
[NEW] added nkey authentication support (adds `userCreds`, `nonceSigner`, `nkey`, `userJWT` connection options.

[CHANGE] renamed examples to tsnats-pub/tsnats-sub, added tsnats-req. examples support creds files.

[CHANGE] server urls are now sanitized if they don't include a protocol or port default protocol or port is added.

### CODE REVIEW

**Comment by aricart** [comment]
squashed!

File: src/protocolhandler.ts [hunk 1]
```diff
@@ -785,6 +786,94 @@ export class ProtocolHandler extends EventEmitter {
...
+            return true;
+        }
+        return false;
+    }
+
+    // returns a regex array - first match is the jwt, second match is the nkey
+    private getUserCreds(jwt = false): RegExpExecArray {
+        if (this.options.userCreds) {
+            let buf = fs.readFileSync(this.options.userCreds);
+            if (buf) {
```
[the remark this answers is not in the document]
Comment (aricart) [reply src/protocolhandler.ts]: nope, null is false.

[the remark this answers is not in the document]
Comment (aricart) [reply src/protocolhandler.ts]: the api call will also toss if there's a problem. Need to check if node-nats handles the exception.

File: README.md [hunk 2]
```diff
@@ -237,6 +237,66 @@ let nc8 = await connect({url: 'nats://token@127.0.0.1:4222'});
 // or token inside the options:
 let nc9 = await connect({url: 'nats://127.0.0.1:4222', token: 'token'});
 ```
+
+### NKey Authentication
+
+NKey authentication returns a public nkey from the client, and signs a challenge.
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: clarified the description.

File: README.md [hunk 3]
```diff
@@ -237,6 +237,66 @@ let nc8 = await connect({url: 'nats://token@127.0.0.1:4222'});
...
+
+NKey authentication returns a public nkey from the client, and signs a challenge. 
+Server authenticates already stores the public key for the client. Client is 
+verified by the signed challenge.
+
+```typescript
+// Seed Keys should be treated as secrets
+const uSeed = "SUAEL6GG2L2HIF7DUGZJGMRUFKXELGGYFMHF76UO2AYBG3K4YLWR3FKC2Q";
+const uPub = "UD6OU4D3CIOGIDZVL4ANXU3NWXOW5DCDE2YPZDBHPBXCVKHSODUA4FKI";
+let nc11 = await connect({url: 'nats://connect.ngs.global', nkey: uPub,
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: changed to tls.

File: README.md [hunk 4]
```diff
@@ -237,6 +237,66 @@ let nc8 = await connect({url: 'nats://token@127.0.0.1:4222'});
...
+        let sk = fromSeed(Buffer.from(uSeed));
+        return sk.sign(Buffer.from(nonce));
+    }
+ });
+```
+
+### JWT Authentication
+
+JWT Authentication returns a JWT to the server and signs a challenge. The JWT
+includes the client's public key and permissions. The server trusts the issuer
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: fixed.

File: README.md [hunk 5]
```diff
@@ -237,6 +237,66 @@ let nc8 = await connect({url: 'nats://token@127.0.0.1:4222'});
...
+    }
+ });
+```
+
+### JWT Authentication
+
+JWT Authentication returns a JWT to the server and signs a challenge. The JWT
+includes the client's public key and permissions. The server trusts the issuer 
+of the JWT which is an operator. The server itself doesn't have a direct knowledge
+of the user.
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: fixed

File: README.md [hunk 6]
```diff
@@ -237,6 +237,66 @@ let nc8 = await connect({url: 'nats://token@127.0.0.1:4222'});
...
+
+```typescript
+// Simples way to connect to a server using JWT and NKeys
+let nc11 = await connect({url: 'nats://connect.ngs.global', userCreds: "/path/to/file.creds"});
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: fixed

File: README.md [hunk 7]
```diff
@@ -407,6 +469,8 @@ The following is the list of connection options and default values.
 | `token`                |                           | Sets a authorization token for a connection
 | `url`                  | `"nats://localhost:4222"` | Connection url
 | `user`                 |                           | Sets the username for a connection
+| `userCreds`            |                           | Path to a properly formatted user credentials file containing the client's JWT and seed key for the client. This property sets a `nonceSigner` automatically.
+| `userJWT`              |                           | A string or a `JwtProvider` function which provides a JWT specifying the client's permissions
```
[the remark this answers is not in the document]
Comment (aricart) [reply README.md]: renamed to JWTProvider.

**Comment by derekcollison** [comment]
LGTM

**Pull request merged**